### DISCUSSION TITLE
Why we create a struct as NFT and not move the created token?

### DISCUSSION THREAD

**Discussion by 0xmaayan** [aptos-labs/aptos-developer-discussions#7] [Questions] [answered]
Hello! not completely sure what is going on here on Aptogotchi contract

At this point we create a token with the account address as the token name https://github.com/aptos-labs/aptogotchi/blob/main/move/sources/aptogotchi.move#L149
but then we create an Aptogotchi struct with the user input name https://github.com/aptos-labs/aptogotchi/blob/main/move/sources/aptogotchi.move#L164

Why do we create a struct and then move it under the token signer? why we dont just move the actual token we created?

**Comment by 0x-j** [comment]
> At this point we create a token with the account address as the token name, but then we create an Aptogotchi struct with the user input name

This implementation is kind of confusing, it would be clearer if we only have 1 name (user passed in), and use it as the token name in token creation, then we don't need to have `name` in `Aptogotchi` struct. But here we use the creator address as token name so we don't need indexer to get the token address. And we add the `name` in `Aptogotchi` struct so we still have a human readable name.

**Reply by 0xmaayan** [reply]
Gotcha. So assuming we do want to use Indexer, we can use the user input name as the token name and remove the use of `Aptogotchi` struct?

**Reply by 0x-j** [reply]
Yes, you can remove `Aptogotchi` struct if you don't need to store additional data about the token, in this example, we want each aptogotchi to have its own data like birthday, etc. So we have this additional struct to store those.

**Comment by 0x-j** [comment, marked as answer]
We did move the created token (the transfer_ref is from the created token) to the creator. 

Another reason Aptogotchi struct is useful is we can store the `extend_ref` of the token (which is an [object](https://aptos.dev/standards/aptos-object) under the hood)  in it. 
So we can add more structs to the token object later (e.g. have an `AptogotchiExtension` struct stores data like location, etc) or update token's name, description and URI. Since once the module is published, we cannot modify existing structs. Only way to store additional data about token is adding new struct to the object, and that requires us to save object's `extend_ref`.

**Comment by 0x-j** [comment]
Here's a [simplified version of the code](https://github.com/aptos-labs/aptogotchi/blob/j/simplify/move/sources/aptogotchi.move#L163) without defining `Aptogotchi` struct. We can use `object_exists<Token>(token_address)` to check existence.
